text/0000-nested-publish.md

```diff
@@ -0,0 +1,183 @@
+
+[lib]
+proc-macro = true
+```
+
+Then you can `cargo publish` from within the parent `foo` directory, and this will create a single `foo` package on `crates.io`, with no `macros` (or `foo-macros`) package visible except when inspecting the source code or in compilation progress messages.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The following changes must be made across Cargo and `crates.io`:
+
+* **Manifest schema**
+    * The Cargo manifest now allows `"nested"` as a value for the `package.publish` key.
+* **`cargo package` &amp; `cargo publish`**
+    * Should refuse to publish a package if that package (not its sub-packages) has `publish = "nested"`.
+    * Exclude/include rules should, upon finding a sub-package, check if it is `publish = "nested"` and not automatically exclude it. Instead, they should treat it like any other subdirectory; in particular, it should be affected by explicitly specified exclude/include rules.
+    * Nested `Cargo.toml`s should be normalized in the same way the root `Cargo.toml` is, if they declare `publish = "nested"`, and not if they do not.
+        * This avoids modifying the publication behavior for existing packages, even if they contain project templates or invoke `cargo` to compile sub-packages to probe the behavior of the compiler.
+        * If the nested `Cargo.toml` has a syntax error such that its `package.publish` value cannot be determined, then if it is depended upon, emit an error; if it is not, emit a warning and do not normalize it.
+* **`crates.io`**
+    * Should allow `path` dependencies that were previously prohibited, at least provided that the named package in fact exists in the `.crate` archive file. The path must not contain any upward traversal (`../`) or other hazardous or non-portable components.
+* **Build process**
+    * Probably some messages will need to be adjusted; currently, `path` dependencies' full paths are always printed in progress messages, but they would be long noise here (`/home/alice/.cargo/registry/src/index.crates.io-6f17d22bba15001f/...`). Perhaps progress for sub-packages could look something like “`Compiling foo/macros v0.1.0`”.
+
+The presence or absence of a `[workspace]` has no effect on the new behavior, just as it has no effect on existing package publication.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* This increases the number of differences between “Cargo package (on disk)” from “Cargo package (that may be published in a registry, or downloaded as a unit)” in a way which may be confusing; it would be good if we have different words for these two entities, but we don't.
+* If Cargo were to add support for multiple libraries per package, that would be largely redundant with this feature.
+* It is not possible to publish a bug fix to a sub-package without republishing the entire parent package.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
```

Comment:
Could you explicitly compare this to
- https://blog.yoshuawuyts.com/inline-crates/
- https://poignardazur.github.io/2023/01/24/subcrates/

And, if it isn't already, incorporate any of the commentary that would be relevant to this proposal (motivation, future possibilities, drawbacks)